src/contrasts.jl

```diff
@@ -0,0 +1,215 @@
+    return throw(
+        ArgumentError("contrast types must be instantiated (use $c() instead of $c)")
+    )
+end
+
+# given an existing ContrastsMatrix, check that all passed levels are present
+# in the contrasts. Note that this behavior is different from the
+# ContrastsMatrix constructor, which requires that the levels be exactly the same.
+function ContrastsMatrix(c::ContrastsMatrix, levels::AbstractVector)
+    if !isempty(setdiff(levels, c.levels))
+        throw(
+            ArgumentError(
+                "there are levels in data that are not in ContrastsMatrix: " *
+                "$(setdiff(levels, c.levels))" *
+                "\n  Data levels: $(levels)" *
+                "\n  Contrast levels: $(c.levels)",
+            ),
+        )
+    end
+    return c
+end
+
+function termnames(C::AbstractContrasts, levels::AbstractVector, baseind::Integer)
+    not_base = [1:(baseind - 1); (baseind + 1):length(levels)]
+    return levels[not_base]
+end
+
+function Base.getindex(contrasts::ContrastsMatrix{C,T}, rowinds, colinds) where {C,T}
+    return getindex(contrasts.matrix, getindex.(Ref(contrasts.invindex), rowinds), colinds)
+end
+
+"""
+    DummyCoding([base[, levels]])
+    DummyCoding(; base=nothing, levels=nothing)
+
+Dummy coding generates one indicator column (1 or 0) for each non-base level.
+
+If `levels` are omitted or `nothing`, they are determined from the data
+by calling the `levels` function on the data when constructing `ContrastsMatrix`.
+If `base` is omitted or `nothing`, the first level is used as the base.
+
+Columns have non-zero mean and are collinear with an intercept column (and
+lower-order columns for interactions) but are orthogonal to each other. In a
+regression model, dummy coding leads to an intercept that is the mean of the
+dependent variable for base level.
+
+Also known as "treatment coding" or "one-hot encoding".
+"""
+mutable struct DummyCoding <: AbstractContrasts
+    base::Any
+    levels::Union{AbstractVector,Nothing}
+end
+## constructor with optional keyword arguments, defaulting to nothing
+function DummyCoding(; base=nothing, levels::Union{AbstractVector,Nothing}=nothing)
+    return DummyCoding(base, levels)
+end
+baselevel(c::DummyCoding) = c.base
+DataAPI.levels(c::DummyCoding) = c.levels
+
+function contrasts_matrix(C::DummyCoding, baseind, n)
+    return Matrix(1.0I, n, n)[:, [1:(baseind - 1); (baseind + 1):n]]
+end
+
+"""
+    FullDummyCoding()
+
+Full-rank dummy coding generates one indicator (1 or 0) column for each level,
+**including** the base level. This is sometimes known as
+[one-hot encoding](https://en.wikipedia.org/wiki/One-hot).
+
+Needed internally for some situations where a categorical variable with ``k``
+levels needs to be converted into ``k`` model matrix columns instead of the
+standard ``k-1``.
+"""
+mutable struct FullDummyCoding <: AbstractContrasts
+    # Dummy contrasts have no base level (since all levels produce a column)
+end
```

Comment:
```suggestion
struct FullDummyCoding <: AbstractContrasts
    # Dummy contrasts have no base level (since all levels produce a column)
end
```
There is nothing to mutate anyway